include/devicetree/pinctrl.h

```diff
@@ -0,0 +1,369 @@
+ * @param idx index into the value of the pinctrl property
+ * @return node identifier for the phandle at index 'idx' in 'pinctrl-'pc_idx''
+ */
+#define DT_PINCTRL_BY_IDX(node_id, pc_idx, idx) \
+	DT_CAT6(node_id, _P_pinctrl_, pc_idx, _IDX_, idx, _PH)
+
+/**
+ * @brief Get a node identifier from a pinctrl-0 property
+ *
+ * This is equivalent to:
+ *
+ *     DT_PINCTRL_BY_IDX(node_id, 0, idx)
+ *
+ * It is provided for convenience since pinctrl-0 is commonly used.
+ *
+ * @param node_id node with a pinctrl-0 property
+ * @param idx index into the pinctrl-0 property
+ * @return node identifier for the phandle at index idx in the pinctrl-0
+ *         property of that node
+ */
+#define DT_PINCTRL_0(node_id, idx) DT_PINCTRL_BY_IDX(node_id, 0, idx)
+
+/**
+ * @brief Get a node identifier for a phandle inside a pinctrl node by name
+ *
+ * Example devicetree fragment:
+ *
+ *     n: node {
+ *             pinctrl-0 = <&foo &bar>;
+ *             pinctrl-1 = <&baz &blub>;
+ *             pinctrl-names = "default", "sleep";
+ *     }
+ *
+ * Example usage:
+ *
+ *     DT_PINCTRL_BY_NAME(DT_NODELABEL(n), default, 1) // DT_NODELABEL(bar)
+ *     DT_PINCTRL_BY_NAME(DT_NODELABEL(n), sleep, 0) // DT_NODELABEL(baz)
+ *
+ * @param node_id node with a named pinctrl property
+ * @param name lowercase-and-underscores pinctrl property name
+ * @param idx index into the value of the named pinctrl property
+ * @return node identifier for the phandle at that index in the pinctrl
+ *         property
+ */
+#define DT_PINCTRL_BY_NAME(node_id, name, idx) \
+	DT_CAT6(node_id, _PINCTRL_NAME_, name, _IDX_, idx, _PH)
+
+/**
+ * @brief Convert a pinctrl name to its corresponding index
+ *
+ * Example devicetree fragment:
+ *
+ *     n: node {
+ *             pinctrl-0 = <&foo &bar>;
+ *             pinctrl-1 = <&baz &blub>;
+ *             pinctrl-names = "default", "sleep";
+ *     }
+ *
+ * Example usage:
+ *
+ *     DT_PINCTRL_NAME_TO_IDX(DT_NODELABEL(n), default) // 0
+ *     DT_PINCTRL_NAME_TO_IDX(DT_NODELABEL(n), sleep)   // 1
+ *
+ * @param node_id node identifier with a named pinctrl property
+ * @param name name of the pinctrl whose index to get
+ * @return integer literal for the index of the pinctrl property with that name
+ */
+#define DT_PINCTRL_NAME_TO_IDX(node_id, name) \
+	DT_CAT4(node_id, _PINCTRL_NAME_, name, _IDX)
+
+/**
+ * @brief Get the number of phandles in a pinctrl property
+ *
+ * Example devicetree fragment:
+ *
+ *     n1: node-1 {
+ *             pinctrl-0 = <&foo &bar>;
+ *     }
+ *
+ *     n2: node-2 {
+ *             pinctrl-0 = <&baz>;
+ *     }
+ *
+ * Example usage:
+ *
+ *     DT_NUM_PINCTRLS_BY_IDX(DT_NODELABEL(n1), 0) // 2
+ *     DT_NUM_PINCTRLS_BY_IDX(DT_NODELABEL(n2), 0) // 1
+ *
+ * @param node_id node identifier with a pinctrl property
+ * @param pc_idx index of the pinctrl property itself
+ * @return number of phandles in the property with that index
+ */
+#define DT_NUM_PINCTRLS_BY_IDX(node_id, pc_idx) \
```

Comment:
Why is this not just `DT_PROP_LEN` of the property?  Not sure we need this macro.

Reply:
> Why is this not just `DT_PROP_LEN` of the property? Not sure we need this macro.

I explained a bit here:

https://github.com/zephyrproject-rtos/zephyr/pull/37621#discussion_r687987955

That's what people are doing now, but they are pasting ``pinctrl_##num`` in their code to do that. Pasting in cpp has "gotchas" that the DT_CATx macros we use under the hood make sure users don't run into.

For example if you do

```c
#define MY_INDEX 1
#define FOO(node_id, pc_idx) DT_PROP_LEN(node_id, pinctrl_ ## pc_idx)

void foo(void)
{
	printk("raw len with pasting: %d\n", FOO(TEST_TEMP, MY_INDEX));
	printk("len with api: %d\n", DT_NUM_PINCTRLS_BY_IDX(TEST_TEMP, MY_INDEX));
}
```

Then this expands to:

```c
void foo(void)
{
 printk("raw len with pasting: %d\n", DT_N_S_test_S_temperature_sensor_P_pinctrl_MY_INDEX_LEN);
 printk("len with api: %d\n", 2);
}
```

Because manual use of `##` in `FOO` misses that you actually have to recursively expand again to get `MY_INDEX` to resolve to an integer literal.

Reply:
question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

Reply:
> question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_REGS
https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_IRQS

Yep, done on purpose for consistency with other APIs like the above.